File: final/evol_strat.py
import random
import math
from typing import List, Tuple

def init_population(pop_size: int, mem_size: int, mem_range: Tuple[float, float], sigma: float, rng: random.Random) -> List[List[float]]:
    population = []
    for _ in range(pop_size):
        member = []
        for _ in range(mem_size):
            gene = rng.uniform(mem_range[0], mem_range[1])
            member.append(gene)
        member.append(sigma)
        population.append(member)
    return population

def evolution_strategy(fitness_fn, mu: int, lambda_: int, mem_size: int, mem_range: Tuple[float, float], sigma: float, tau: float, max_gens: int, rng: random.Random) -> List[List[float]]:
    population = init_population(mu, mem_size, mem_range, sigma, rng)
    cumulative_evals = 0

    for generation_number in range(1, max_gens + 1):
        fitnesses = [fitness_fn.fitness(member[:mem_size]) for member in population]
        cumulative_evals += mu

        offspring = []
        for _ in range(lambda_):
            # Select parent using tournament
            candidates = rng.sample(range(mu), 2)
            parent_idx = max(candidates, key=lambda i: fitnesses[i])
            parent = population[parent_idx]

            child = []
            genes = parent[:mem_size]
            sigma_val = parent[mem_size]
            for gene in genes:
                mutation = rng.gauss(0.0, sigma_val)
                mutated_gene = gene + mutation
                child.append(mutated_gene)
            # Mutate sigma
            sigma_mutation = rng.gauss(0.0, 1.0)
            new_sigma = sigma_val * math.exp(tau * sigma_mutation)
            child.append(new_sigma)
            offspring.append(child)

        offspring_fitnesses = [fitness_fn.fitness(member[:mem_size]) for member in offspring]
        cumulative_evals += lambda_

        max_fitness = max(fitnesses)
        average = sum(fitnesses) / mu
        diversity = 0.0
        for i in range(mu):
            for j in range(i + 1, mu):
                dist = math.sqrt(sum((population[i][k] - population[j][k])**2 for k in range(mem_size)))
                if dist > diversity:
                    diversity = dist
        print(f"Himmelblau ES {mu} {lambda_} {tau} 0.0 {generation_number} {cumulative_evals} {max_fitness} {average} {diversity}")

        if average > 0.99:
            break

        # Select best mu from offspring
        indexed = [(f, i) for i, f in enumerate(offspring_fitnesses)]
        indexed.sort(key=lambda x: x[0], reverse=True)
        selected_indices = [i for _, i in indexed[:mu]]

        new_population = [offspring[idx] for idx in selected_indices]
        population = new_population

    return population

File: HW3/src/evol_strat.rs
use crate::fitness::Fitness;
use rand::prelude::*;
use rand_chacha::ChaCha8Rng;

pub struct ESParameters {
    pub mu: usize,
    pub lambda: usize,
    pub mem_size: usize,
    pub mem_range: (f64, f64),
    pub sigma: f64,
    pub tau: f64,
    pub max_gens: usize,
}

fn init_population(
    params: &ESParameters,
    rng: &mut ChaCha8Rng,
) -> Vec<Vec<f64>> {
    let mut population = Vec::new();
    for _ in 0..params.mu {
        let mut member: Vec<f64> = Vec::new();
        for _ in 0..params.mem_size {
            let gene: f64 = rng.random_range(params.mem_range.0..params.mem_range.1);
            member.push(gene);
        }
        member.push(params.sigma);
        population.push(member);
    }
    population
}

pub fn evolution_strategy<F: Fitness>(
    fitness_fn: &F,
    params: &ESParameters,
    rng: &mut ChaCha8Rng,
) -> Vec<Vec<f64>> {
    // Initialize population
    let mut population = init_population(params, rng);
    let mut cumulative_evals = 0;

    for generation_number in 1..=params.max_gens {
        // Evaluate fitness of current population
        let fitnesses: Vec<f64> = population
            .iter()
            .map(|member| fitness_fn.fitness(&member[0..params.mem_size]))
            .collect();
        cumulative_evals += params.mu;

        // Create lambda offspring
        let mut offspring = Vec::new();
        for _ in 0..params.lambda {
            // Select a parent using tournament selection
            let parent_idx = (0..params.mu)
                .choose_multiple(rng, 2)
                .into_iter()
                .max_by(|&i, &j| fitnesses[i].partial_cmp(&fitnesses[j]).unwrap())
                .unwrap();
            let parent = &population[parent_idx];

            // Mutate the parent to create an offspring
            let mut child = Vec::new();
            let genes = &parent[0..params.mem_size];
            let sigma_val = parent[params.mem_size];
            for &gene in genes {
                let mutation: f64 =
                    rng.sample::<f64, _>(rand_distr::Normal::new(0.0, sigma_val).unwrap());
                let mutated_gene = gene + mutation;
                child.push(mutated_gene);
            }
            // Mutate sigma
            let sigma_mutation: f64 =
                rng.sample::<f64, _>(rand_distr::Normal::new(0.0, 1.0).unwrap());
            let new_sigma = sigma_val * (params.tau * sigma_mutation).exp();
            child.push(new_sigma);
            offspring.push(child);
        }

        // Evaluate fitness of offspring
        let offspring_fitnesses: Vec<f64> = offspring
            .iter()
            .map(|member| fitness_fn.fitness(&member[0..params.mem_size]))
            .collect();
        cumulative_evals += params.lambda;

        let max_fitness = fitnesses.iter().fold(f64::NEG_INFINITY, |a, &b| a.max(b));
        let sum: f64 = fitnesses.iter().sum();
        let average = sum / params.mu as f64;
        let mut diversity = 0.0;
        for i in 0..params.mu {
            for j in (i + 1)..params.mu {
                let dist = (0..params.mem_size)
                    .map(|k| (population[i][k] - population[j][k]).powi(2))
                    .sum::<f64>()
                    .sqrt();
                if dist > diversity {
                    diversity = dist;
                }
            }
        }
        println!(
            "Dejong Rosenbrock ES {} {} {} 0.0 {} {} {} {} {}",
            params.mu, params.lambda, params.tau, generation_number, cumulative_evals, max_fitness, average, diversity
        );

        // Early stopping if average fitness exceeds threshold
        if average > 0.99 {
            break;
        }

        // Select the best mu from lambda offspring
        let mut indexed: Vec<(f64, usize)> = offspring_fitnesses
            .iter()
            .enumerate()
            .map(|(i, &f)| (f, i))
            .collect();
        indexed.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap()); // descending
        let selected_indices: Vec<usize> = indexed.into_iter().take(params.mu).map(|(_, i)| i).collect();

        let mut new_population = Vec::new();
        for &idx in &selected_indices {
            new_population.push(offspring[idx].clone());
        }

        population = new_population;
    }

    population
}
